### PR TITLE
[AKS] `az aks install-desktop`: Add AKS Desktop installation

### PR DESCRIPTION
**Pull request by gambtho**
### 🤖 PR Validation — ⚠️ Review suggested

| Breaking Changes | Tests |
| :--: | :--: |
| ⚠️ None | ️✔️ 130/130 |

<details>
<summary>⚠️AzureCLI-BreakingChangeTest</summary>

><details>
> <summary>⚠️acs</summary>
>
>|rule|cmd_name|rule_message|suggest_message|
>|---|---|---|---|
>|⚠️ [1001 - CmdAdd](https://github.com/Azure/azure-cli/blob/dev/doc/breaking_change_rules/1001.md)|aks install-desktop|cmd `aks install-desktop` added||
>
></details>
</details>

## Summary

- add `az aks install-desktop` with latest-stable installation by default and optional `--version`
- select exact AKS Desktop release assets for Windows, macOS, and Linux architectures
- stream downloads over HTTPS, reject insecure redirects, verify GitHub-published SHA-256 digests, and clean temporary files
- launch native installers without shell execution or privilege escalation; install archive-only Linux builds under the current user's profile
- add help examples, unit regressions, and five offline command-level scenarios covering argument parsing, authenticated metadata access, integrity verification, safe extraction, and cleanup

## Validation

- Combined `test_custom` and offline CLI scenario suites: 105 passed, 1 pre-existing skip, 125 subtests passed.
- Archive extraction regressions pass in isolation on actual Python 3.10.11, 3.11.3, 3.12.3, 3.14.3 and 3.14.6. The checksum-verified ARM64 release was extracted successfully on both pre-backport runtimes without application execution.
- A real `xdg-open` handoff with an isolated harmless handler confirms the retained installer remains readable after the command returns.
- Modified Python files compile and pass Python 3.10 syntax checks; source Pylint, changed-line pycodestyle, and `git diff --check` pass.
- Source CLI `az aks install-desktop -h` renders the arguments, examples and installer-retention guidance.
- The repository PR-title/content checker passes after correcting the title format.
- `/polish` and CodeRabbit reviews completed; review findings were addressed.

### CODE REVIEW

File: src/azure-cli/azure/cli/command_modules/acs/custom.py [hunk 1]
```diff
@@ -2385,6 +2387,212 @@ def k8s_install_cli(cmd, client_version='latest', install_location=None, base_sr
...
+
+
+def _launch_aks_desktop_installer(installer_path, system, version):
+    try:
+        if system == 'win':
+            subprocess.run([installer_path], check=True)
+        elif system == 'mac':
+            subprocess.run(['open', installer_path], check=True)
+        elif installer_path.endswith('.deb'):
+            subprocess.run(['xdg-open', installer_path], check=True)
```
[the remark this answers is not in the document]
Comment (gambtho) [reply src/azure-cli/azure/cli/command_modules/acs/custom.py:2560]: Addressed in e9a15c4b7550c6236e8b0f271916ddeb58add004. Successful macOS/DEB GUI handoffs now retain the verified installer in a unique directory under the Azure CLI configuration directory and print its location and cleanup instructions. Download/launch failures and synchronous installers still clean up. Regression tests cover delayed consumption and repeated invocations; a real xdg-open test with an isolated harmless handler also confirmed the file remains readable after the command returns.

File: src/azure-cli/azure/cli/command_modules/acs/custom.py [hunk 2]
```diff
@@ -2385,6 +2387,212 @@ def k8s_install_cli(cmd, client_version='latest', install_location=None, base_sr
...
+                if os.path.commonpath((destination_path, member_path)) != destination_path:
+                    raise FileOperationError(
+                        'The AKS Desktop archive contains an unsafe path.')
+                if member.issym() or member.islnk():
+                    link_path = os.path.realpath(os.path.join(
+                        os.path.dirname(member_path), member.linkname))
+                    if os.path.commonpath((destination_path, link_path)) != destination_path:
+                        raise FileOperationError(
+                            'The AKS Desktop archive contains an unsafe link.')
+            archive.extractall(destination)  # nosec B202 - member paths are validated above
```
[the remark this answers is not in the document]
Comment (gambtho) [reply src/azure-cli/azure/cli/command_modules/acs/custom.py]: Addressed in 5f25cf0544. Extraction now explicitly uses tarfile’s data filter, so path/link checks run against the filesystem state left by earlier members. Runtimes without the safe filter fail closed with an actionable Python-update error. I also reproduced the trailing-separator symlink bypass on Python 3.14.3 (CPython gh-149486) and added an explicit rejection for those malformed link names. Regression coverage includes the reported symlink escape, hardlink traversal, direct traversal, FIFO members, the older-runtime bypass, and valid contained links. Isolated archive tests pass on Python 3.12.3, 3.14.3 and 3.14.6; the full test_custom suite passes (80 run, 1 skipped).

File: src/azure-cli/azure/cli/command_modules/acs/custom.py [hunk 3]
```diff
@@ -2385,6 +2387,232 @@ def k8s_install_cli(cmd, client_version='latest', install_location=None, base_sr
...
+                "AKS Desktop version '{}' is invalid. Specify a semantic version such as '0.9.1'."
+                .format(version))
+        requested_version = version
+        release_url = '{}/tags/v{}'.format(_AKS_DESKTOP_RELEASES_API, version)
+    else:
+        release_url = _AKS_DESKTOP_RELEASES_API + '/latest'
+
+    logger.warning('Getting AKS Desktop release metadata from "%s".', release_url)
+    try:
+        release = json.loads(_urlopen_read(release_url))
```
Comment (FumingZhang) [inline src/azure-cli/azure/cli/command_modules/acs/custom.py]: [P2] Make the GitHub metadata request recover from rate limiting. This call is unauthenticated, so it gets GitHub's shared 60-requests/hour quota; moreover, the error's recommendation to pass `--version` cannot recover because that path simply calls another unauthenticated REST endpoint (`/releases/tags/...`) and receives the same 403/429. This module's adjacent kubelogin flow already recognizes this failure mode and offers authentication/fallback. Please provide an authenticated or non-API metadata path here and make the recommendation actually bypass the exhausted quota.

File: src/azure-cli/azure/cli/command_modules/acs/custom.py [hunk 4]
```diff
@@ -2385,6 +2387,232 @@ def k8s_install_cli(cmd, client_version='latest', install_location=None, base_sr
...
+
+    if actual_digest.hexdigest().lower() != digest.split(':', 1)[1].lower():
+        raise ClientRequestError(
+            'The downloaded AKS Desktop artifact did not match its published SHA-256 digest.')
+
+
+def _extract_aks_desktop_archive(archive_path, destination):
+    if getattr(tarfile, 'data_filter', None) is None:
+        raise FileOperationError(
+            'Safe AKS Desktop archive extraction requires an updated Python with tarfile.data_filter support.')
```
Comment (FumingZhang) [inline src/azure-cli/azure/cli/command_modules/acs/custom.py]: [P2] Preserve archive installs on all supported Python runtimes. Azure CLI declares `python_requires='>=3.10.0'`, but `tarfile.data_filter` is absent from upstream CPython 3.10.0-3.10.11 and 3.11.0-3.11.3. On those valid runtimes every archive-only target (Linux ARM, non-Debian x64, or x64 without `xdg-open`) deterministically fails before extraction. Please use an equivalent safe fallback on pre-backport runtimes (or otherwise align the command/runtime support contract) rather than making the new command unusable for that supported subset.

**Review by FumingZhang — COMMENTED** [review]
please consider adding a scenario test case to help validate the change

File: src/azure-cli/azure/cli/command_modules/acs/custom.py (hunk 3 above)
Comment (gambtho) [reply src/azure-cli/azure/cli/command_modules/acs/custom.py]: Addressed in 69b61dfe66, with command-level coverage in acb91b9388. `--gh-token` now authenticates both latest and explicit-version metadata requests through the existing transport helper. Anonymous 429/evidenced-403 failures recommend that option or waiting for reset; authenticated throttling recommends checking the token quota/waiting, not changing `--version`. Other HTTP/network/JSON failures keep their context. The token is not passed to installer/CDN downloads, and malformed CR/LF-containing tokens are rejected without exposing their value. Unit and offline ScenarioTest coverage verifies the actual CLI argument routing and request headers.

File: src/azure-cli/azure/cli/command_modules/acs/custom.py (hunk 4 above)
Comment (gambtho) [reply src/azure-cli/azure/cli/command_modules/acs/custom.py]: Addressed in 02d4f566df, with final cross-version test coverage in acb91b9388. Runtimes without tarfile filters now use a private compatibility extractor that checks current filesystem containment at each operation, streams regular files explicitly, validates symlink/hardlink targets, strips unsafe permissions, and rejects special members and ambiguous deferred hardlinks. The native data-filter path and trailing-separator protection remain. No dependency or Python-minimum changes. Actual archive regression tests pass on Python 3.10.11 and 3.11.3 (both genuinely lack filter APIs), plus 3.12.3, 3.14.3 and 3.14.6. The checksum-verified v0.9.1 ARM64 release was also extracted into disposable directories on both legacy runtimes: all 37,788 entries and 1,049 symlinks, without executing the application.

**Comment by gambtho** [comment]
@FumingZhang Added the requested command-level scenario coverage in acb91b9388: `tests/latest/test_aks_install_desktop.py` uses `ScenarioTest.cmd` with real registration, argument parsing, metadata handling, hashing, extraction and cleanup. Five offline cases cover pinned/latest versions, metadata-only authentication, legacy extraction, invalid/empty versions, and digest mismatch. HTTP and process boundaries are stubbed; no application is executed and no Azure credentials are required.

Also addressed the earlier validation/test feedback: null/non-string digests fail cleanly, empty versions no longer select latest, and cleanup fixtures write/read actual files.

File: src/azure-cli/azure/cli/command_modules/acs/custom.py [hunk 5]
```diff
@@ -2385,6 +2387,328 @@ def k8s_install_cli(cmd, client_version='latest', install_location=None, base_sr
...
+        ('linux', 'armv7l'): ['aks-desktop-{}-linux-armv7l.tar.gz'.format(version)],
+    }.get((system, arch), [])
+
+    if (system, arch) == ('linux', 'x64'):
+        try:
+            os_release = platform.freedesktop_os_release()
+        except OSError:
+            os_release = {}
+        distro_ids = {os_release.get('ID'), *os_release.get('ID_LIKE', '').split()}
+        if distro_ids.intersection(('debian', 'ubuntu')) and shutil.which('xdg-open'):
```
Comment (FumingZhang) [inline src/azure-cli/azure/cli/command_modules/acs/custom.py:2487]: [P2] Do not treat the presence of `xdg-open` as proof that a DEB handler is usable. On headless or SSH Ubuntu systems, `xdg-open` is commonly installed but exits nonzero because there is no desktop session or package handler. The command then selects the `.deb`, deletes it after the failed handoff, and never uses the available portable archive. I reproduced this on Ubuntu 24.04: asset selection chose the DEB and `xdg-open` returned 1. Please gate this choice on a usable graphical session/handler or fall back to the tarball when the DEB handoff fails.

File: src/azure-cli/azure/cli/command_modules/acs/custom.py [hunk 6]
```diff
@@ -2385,6 +2387,328 @@ def k8s_install_cli(cmd, client_version='latest', install_location=None, base_sr
...
+        if system == 'win':
+            subprocess.run([installer_path], check=True)
+        elif system == 'mac':
+            subprocess.run(['open', installer_path], check=True)
+        elif installer_path.endswith('.deb'):
+            subprocess.run(['xdg-open', installer_path], check=True)
+        else:
+            install_dir = os.path.join(
+                os.path.expanduser('~'), '.local', 'share', 'aks-desktop', version)
+            _extract_aks_desktop_archive(installer_path, install_dir)
```
Comment (FumingZhang) [inline src/azure-cli/azure/cli/command_modules/acs/custom.py]: [P2] Extract into a clean or atomically replaced version directory. This overlays the archive onto any existing `~/.local/share/aks-desktop/<version>` contents. Stale files therefore survive reinstalls, and a stale file named `aks-desktop` can satisfy the subsequent `os.walk` search even when the newly verified archive contains no executable. I reproduced both the native and compatibility paths launching a pre-existing `old/aks-desktop` after extracting such an archive. Please extract into a sibling temporary directory, validate the expected executable there, and atomically replace the version directory; also remove the temporary directory on failure.

File: src/azure-cli/azure/cli/command_modules/acs/tests/latest/test_aks_install_desktop.py [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+        self.metadata.assert_called_once()
+        request = self.metadata.call_args.args[0]
+        self.assertEqual(request.full_url, 'https://api.github.com/repos/Azure/aks-desktop/releases/' + endpoint)
+        self.assertEqual(request.get_header('Authorization'), 'Bearer ' + token if token else None)
+        self.opener.open.assert_called_once()
+        executable = self.home / '.local' / 'share' / 'aks-desktop' / '0.9.1' / 'app' / 'aks-desktop'
+        self.assertEqual(executable.read_bytes(), b'fixture')
+        if os.name != 'nt':
+            self.assertEqual(stat.S_IMODE(executable.stat().st_mode), 0o755)
+        self.popen.assert_called_once_with([str(executable)])
```
Comment (FumingZhang) [inline src/azure-cli/azure/cli/command_modules/acs/tests/latest/test_aks_install_desktop.py]: [P2] Assert only the AKS Desktop launch instead of every `Popen` call. The test globally patches `subprocess.Popen` for the entire CLI invocation, so unrelated runtime probes are counted too. In an isolated Linux run, the calls were `('/usr/sbin/ip', 'link')`, `('/usr/sbin/ifconfig',)`, and then the expected desktop executable, causing this assertion to fail even though installation succeeded. Later cases can pass after interface discovery is cached, making the result order-dependent. Please filter `call_args_list` for the expected executable or patch the launcher at a narrower boundary.

File: src/azure-cli/azure/cli/command_modules/acs/_params.py [hunk 8]
```diff
@@ -1025,6 +1025,11 @@ def load_arguments(self, _):
         c.argument('kubelogin_base_src_url', options_list=['--kubelogin-base-src-url', '-l'], help='Base download source URL for kubelogin releases.')
         c.argument('gh_token', help='GitHub authentication token used when downloading kubelogin binaries from GitHub releases. Supplying a token helps avoid GitHub API rate limits.')
 
+    with self.argument_context('aks install-desktop') as c:
+        c.argument('version', help='Version of AKS Desktop to install. By default, the latest stable version is installed.')
+        c.argument('gh_token', help='GitHub authentication token used to retrieve AKS Desktop release metadata. '
```
Comment (FumingZhang) [inline src/azure-cli/azure/cli/command_modules/acs/_params.py:1030]: [P3] Avoid making argv the recommended way to supply the GitHub token. Azure CLI logs raw command arguments before this command can sanitize them; I reproduced `az aks install-desktop --gh-token <token> --debug` emitting the token in `DEBUG: ... Command arguments`. It is also exposed through shell history and the process list. Since the help now recommends a token for rate-limit recovery, please support a standard environment variable such as `GH_TOKEN` or another non-argv input and recommend that path for credentials. The adjacent `aks install-cli --gh-token` has the same pre-existing concern, so this may warrant a shared follow-up.

File: src/azure-cli/azure/cli/command_modules/acs/custom.py (hunk 5 above)
Comment (gambtho) [reply src/azure-cli/azure/cli/command_modules/acs/custom.py:2487]: Addressed in 586780382f. Linux x64 now prefers the DEB only with a graphical-session indicator (DISPLAY/WAYLAND_DISPLAY), compatible distro and xdg-open. Those indicators are not treated as proof of success: a failed DEB handoff retries the portable archive from the same fetched release metadata, after cleaning the failed DEB. Download/checksum failures remain terminal; the archive cannot retry indefinitely. Regression coverage includes headless selection, handoff failures, missing fallback and checksum-failure boundaries.

File: src/azure-cli/azure/cli/command_modules/acs/custom.py (hunk 6 above)
Comment (gambtho) [reply src/azure-cli/azure/cli/command_modules/acs/custom.py]: Addressed in 586780382f, with cleanup hardening in 1d26969bf6. Archives now extract into a fresh sibling staging directory; executable discovery and chmod happen there before the existing installation is touched. Successful publication replaces the whole version tree, so stale files cannot survive or satisfy validation. Extraction/validation/publication failures preserve the previous installation; publication rollback also handles interruption, and failed rollback retains/reports the recovery backup. A nonempty directory needs a backup rename before os.replace publication, so the version path is briefly absent—not a single atomic exchange. Post-publication backup cleanup failures warn with a manual-cleanup path rather than preventing launch. Native/compatibility regressions pass on Python 3.10.11, 3.11.3, 3.12.3, 3.14.3 and 3.14.6; the official ARM64 archive also passed an actual staged reinstall without application execution.

File: src/azure-cli/azure/cli/command_modules/acs/tests/latest/test_aks_install_desktop.py (hunk 7 above)
Comment (gambtho) [reply src/azure-cli/azure/cli/command_modules/acs/tests/latest/test_aks_install_desktop.py]: Addressed in 586780382f. Both success and failure assertions now inspect only AKS Desktop launches, while retaining exact argv and single-launch checks. The fixture deliberately records ip/ifconfig probe calls in every case, so cached interface discovery can no longer hide this regression. Those injected calls reproduced the previous failures before the fix; all seven scenarios now pass, including in reversed order. Actual subprocess execution remains mocked.

File: src/azure-cli/azure/cli/command_modules/acs/_params.py (hunk 8 above)
Comment (gambtho) [reply src/azure-cli/azure/cli/command_modules/acs/_params.py:1030]: Addressed in 586780382f. install-desktop now accepts GH_TOKEN when --gh-token is omitted; help and rate-limit recovery recommend the environment variable rather than argv. The existing explicit flag still takes precedence for compatibility, with a warning about history/process-list/debug-log exposure. A real CLI --debug scenario verifies authenticated metadata, no Authorization header on the artifact request, and absence of the environment token from captured debug logs (including the actual Command arguments record). Malformed environment tokens use the same secret-free validation. This does not claim to redact explicitly supplied argv tokens or change the adjacent install-cli/shared logger behavior; that remains separate scope.

**Comment by gambtho** [comment]
Addressed the four new review threads in 586780382f and 1d26969bf6. Local unit/scenario verification: 116 passed, 1 existing skip; archive/install regressions also pass on Python 3.10–3.14.

@yonzhan Could you trigger `/azp run` for the updated head (`1d26969bf6`)? My previous attempt was rejected for insufficient privileges.

**Review by FumingZhang — APPROVED** [review]
